src/com/xilinx/rapidwright/rwroute/RWRoute.java

```diff
@@ -0,0 +1,1843 @@
+			if(!this.targetReached() && !successRoute) {
+				Routable rnode = this.queue.poll();
+				this.nodesPopped++;
+				
+				this.setChildrenOfRnode(rnode);
+				this.exploreAndExpand(rnode, connection, shareWeight, rnodeCostWeight,
+						wlWeight, estWlWeight, dlyWeight, estDlyWeight);
+			}else {
+				successRoute = true;
+				break;
+			}
+		}
+		
+		if(successRoute) {
+			this.finishRouteConnection(connection);
+			connection.getSink().setRouted(true);
+			if(config.isTimingDriven()) connection.updateRouteDelay();	
+		}else {
+			connection.getSink().setRouted(false);
+			connection.getSinkRnode().setTarget(false);
+			this.resetExpansion();
+			System.out.printf("CRITICAL WARNING: Unroutable connection in iteration #%d\n", this.routeIteration);
+			System.out.println("                 " + connection);
+			this.handleUnroutableConnection(connection);
+		}
+	}
+	
+	/**
+	 * Deals with a failed connection by possible output pin swapping and unrouting preserved nets if the router is in the soft preserve mode.
+	 * @param connection The failed connection.
+	 */
+	private void handleUnroutableConnection(Connection connection) {
+		if (this.routeIteration == 1) {		
+			boolean hasAltOutput = this.swapOutputPin(connection);
+			if(!hasAltOutput && this.config.isSoftPreserve()) {
+				this.unrouteReservedNetsToReleaseResources(connection);
+			}
+		} else if(this.routeIteration == 2) {
+			if (this.config.isSoftPreserve()) this.unrouteReservedNetsToReleaseResources(connection);
+		}
+	}
+	
+	/**
+	 * Swaps the output pin of a connection, if its net has an alternative output pin.
+	 * @param connection The connection in question.
+	 * @return true, if the output pin has been swapped.
+	 */
+	private boolean swapOutputPin(Connection connection) {	
+		NetWrapper np = connection.getNetWrapper();
+		Net n = np.getNet();
+		
+		SitePinInst altSource = DesignTools.getLegalAlternativeOutputPin(n);
+		if(altSource == null) {
+			System.out.println("INFO: No alternative source to swap");	
+			return false;
+		}
+		
+		System.out.println("INFO: Swap source from " + n.getSource() + " to " + altSource + "\n");
+		
+		n.replaceSource(altSource);
+		n.setSource(altSource);
+		n.setAlternateSource(connection.getSource());
+		DesignTools.routeAlternativeOutputSitePin(n, altSource);
+		np.setSourceChanged(true, connection.getSource());
+		
+		Node sourceINTNode = RouterHelper.projectOutputPinToINTNode(altSource);
+		Routable sourceR = this.createAddRoutableNode(this.rnodeId, altSource, sourceINTNode, RoutableType.PINFEED_O);;
+		for(Connection c : np.getConnection()) {
+			c.setSource(altSource);
+			c.setSourceRnode(sourceR);
+		}
+			
+		return true;
+	}
+	
+	/**
+	 * Unroutes preserved nets to release routing resource to resolve congestion that blocks the routablity of a connection.
+	 * @param connection The connection in question.
+	 * @return The number of unrouted nets.
+	 */
+	private int unrouteReservedNetsToReleaseResources(Connection connection) {
```

Comment:
Is my understanding correct that this just unroutes reserved nets that are occupying resources immediately downhill of the source, and immediately uphill of the sink? If the congestion is outside of these immediate vicinities, then may not help?

Also, it looks like you're unrouting the whole reserved net. Have you considered unrouting just the associated routing branch of the reserved net (assuming it has more than one sink) in order to preserve as much as possible?

Reply:
That is a correct understanding. If the congestion or blockage if outside of these immediate vicinities, enlarging bounding boxes is there to help by allowing enough routing detours.

Unrouting preserved nets is intended to be used for the soft preserve mode. The partial unrouting is not explored yet. (PS: there could be multiple associated branches of a preserved net to a node).

Reply:
Enlarging the bounding box may not be sufficient in all cases. Let's say net 1 is trying to route. If there is at least one preserved net using all of the resources that net 1 needs to successfully route, and those preserved nets do not appear in the net 1's source and sink vicinities, then they will never be unrouted (in soft preserve mode).

My intuition right now is that the root of this stems from the fact that one can *never* use reserved nodes as they simply don't appear in the routing graph. It may be beneficial to allow nets that really need to use a reserved resource to use it (heavily discouraged by setting a very high cost) so that we can examine exactly which reserved nets need to be unrouted.

It also appears that all reserved nets in the vicinity of the source and sink are fully unrouted, even if they are not needed. The approach above might be able to help this. I'm experimenting with this router for constrained scenarios (restricting routing inside a PBlock like `CONTAIN_ROUTING` does) as well as trying to encourage the router to rip up as little as possible.

Regarding unrouting a branch of the net, yes, a branch may lead to multiple sinks depending on if it's a thick or thin branch, but I am suggesting that a subset of all sinks be unrouted where possible (i.e. when the conflicting node does not lie on the trunk) instead of the entire net -- which matches my objective of ripping up as little as possible.

I'm not suggesting you fix this now as it could be a big change on top of what is already a big PR, but perhaps update the comment to indicate this limitation?

Reply:
I understand the point and appreciate the discussion. Updating the method javadoc to indicate the limitation has been done. 

If you would like to experiment with the partial routing feature and need the router to rip up as little as possible, the current preserving mechanism would need to be changed. For instance, it makes all used routing resources of the preserved nets invisible during the routing process, which impedes the negotiation among routing-needing nets and preserved nets.